packages/server/movement.ts

```diff
@@ -0,0 +1,242 @@
+
+        // see if current direction is blocked
+        const o = obstacles.find(o => overlap(o, [d, d]));
+        // if yes, aim for its edge
+        // TODO consider own size instead of EXTRA_TURN
+        if (o) {
+            const EXTRA_TURN = 0.1;
+            return Math.abs(o[0] - d) < Math.abs(o[1] - d)
+                    ? o[0]-EXTRA_TURN
+                    : o[1]+EXTRA_TURN;
+        } else {
+            return d;
+        }
+    })();
+    
+    let separation = {x:0, y:0};
+    for (const u of otherUnitsNearby) {
+        const MAX_LOCAL_SEPARATION_FORCE = 2;
+        const SEPARATION_OVERCOMP = 1.05; // overcompensation for separation distance
+
+        let localSeparation = V.difference(currentPos, u.position);    
+        const distance = V.magnitude(localSeparation);
+
+        // TODO real sizes or buildings which are square
+        const myOwnRadius = 1;
+        const otherUnitRadius = 1;
+
+        if (distance > (myOwnRadius + otherUnitRadius) * SEPARATION_OVERCOMP)
+            continue;
+
+        // the force gets stronger the closer it is
+        const distanceFactor = distance > 0.00001 ? 0.2 / distance : MAX_LOCAL_SEPARATION_FORCE; // avoid zero distance issues
+        localSeparation.x *= distanceFactor;
+        localSeparation.y *= distanceFactor;
+
+        // clamp the local force to avoid very high impulses at close passes
+        localSeparation = V.clamp(localSeparation, MAX_LOCAL_SEPARATION_FORCE);
+        separation = V.sum(separation, localSeparation);
+
+        // push other unit apart (but only if it can move)
+        
+        //if (u.components.find(c => c.type === 'Mover')) {
+        //    u.position.x -= localSeparation.x;
+        //    u.position.y -= localSeparation.y;
+        //}
+    }
+
+    // limit maximum    
+    const MAX_SEPARATION_FORCE = 3;
+    separation = V.clamp(separation, MAX_SEPARATION_FORCE);
+
+    // push off of terrain
+    const terrainAvoidance = {x: 0, y:0};
+    {
+        const terrainNearby =
+            allTilesInfluenced
+            .map(t => gm.tiles[explode(t)] !== 0 ? t : undefined)
+            .filter(notEmpty);
+        unit.debug.terrainNearby = terrainNearby;
+
+        for (const t of terrainNearby) {
+            const diff = V.difference(currentPos, t);
+            const distance = V.magnitude(diff);
+
+            // TODO radius etc
+            if (distance > 1.5) {
+                continue;
+            }
+
+            // convert to unit vector
+            diff.x /= distance;
+            diff.y /= distance;
+            V.vecAdd(terrainAvoidance, diff);
+        }
+    }
+    unit.debug.terrainAvoidance = terrainAvoidance;
+
+    let velocity = {
+        x: Math.cos(nearestGapAngle) * 1,
+        y: -Math.sin(nearestGapAngle) * 1
+    };
+
+    V.vecAdd(velocity, separation);
+    V.vecAdd(velocity, terrainAvoidance);
+    
+    return velocity;
+}
+
+const wrap360 = (x: number) => {

```

Comment:
This looks very useful, I'm wondering if it's possible to make it more generic so that big values like 1080 deg are normalized too - maybe it's possible to achieve if you use some combination of `%` and `Math.abs`

Reply:
I think I'm wrong about `%` but I feel like there's some nice clean generic way to solve this problem

Reply:
Yeah, that was laziness on my part, good catch. The easiest way is to replace `if` with `while` i guess

Reply:
`while` looks very decent here, I didn't though of that